### ISSUE TITLE
Collection like types not map items

### ISSUE THREAD

**Issue by sky-code** [AutoMapper/AutoMapper#1536]
AutoMapper not map all items in types derived from List<T>
I Mean something like that:

``` csharp
public class PagedList<T> : List<T>, IPagedList<T>
{
        public int PageCount { get; set; }
        public int TotalItemCount { get; set; }
        public int PageNumber { get; set; }
        // ..
}
```

Type has indexer + some other properties
In that case mapping behaviour differs from version to version
Real use case, we need to map `PagedList<TEntity>` to `PagedList<TViewModel>`, create mapping for that:

``` csharp
config.CreateMap(typeof(PagedList<>), typeof(PagedList<>))
```

and when we map some object with AutoMapper 5.0.2, we get mapped only type properties but not collection items, result list will be empty.
In AutoMapper 4.2.1 behavior for that differs, we get list items mapped but not properties.
How to map list items and properties?
Is this a bug?
Different behavior on each version of AutoMapper and lack of documentation for that case sometimes can be real problem.

Documentation for `Lists and arrays` missing case when you have `source` IEnumerable type with properties and `destination` Collection type with properties, what behavior has to be in such situation?
Add documentation of AutoMapper behavior when your try map `Collection` like type with properties.

**Comment by sky-code** [comment]
Here is example code
```csharp
class Program
    {
        static void Main()
        {
            Mapper.Initialize(config =>
            {
                config.CreateMap<UserEntity, UserViewModel>()
                    .ForMember(d => d.UserId, c => c.MapFrom(s => s.Id))
                    .ForMember(d => d.UserName, c => c.MapFrom(s => s.Name));

                config.CreateMap(typeof(PagedList<>), typeof(PagedList<>))
                    .IncludeBase(typeof(List<>), typeof(List<>));

                config.CreateMap<PagedList<UserEntity>, MainViewModel>()
                    .ForMember(d => d.Users, c => c.MapFrom(s => s))
                    .ForMember(d => d.SomeOtherData, c => c.UseValue("any"));
            });

            var source = new PagedList<UserEntity>
            {
                new UserEntity {Id = 1, Name = "first"},
                new UserEntity {Id = 2, Name = "second"},
                new UserEntity {Id = 3, Name = "last"}
            };
            source.PageSize = 2;
            source.TotalItemCount = 3;

            var model = new MainViewModel();
            Mapper.Map(source, model);

            Console.WriteLine(model.Users?.Count);
        }
    }

    public class UserEntity
    {
        public int Id { get; set; }

        public string Name { get; set; }
    }

    public class UserViewModel
    {
        public int UserId { get; set; }

        public string UserName { get; set; }
    }

    public class MainViewModel
    {
        public PagedList<UserViewModel> Users { get; set; }

        public string SomeOtherData { get; set; }
    }

    public class PagedList<T> : List<T>, IPagedList<T>
    {
        public int PageCount { get; set; }
        public int TotalItemCount { get; set; }
        public int PageNumber { get; set; }
        public int PageSize { get; set; }
        public bool HasPreviousPage { get; set; }
        public bool HasNextPage { get; set; }
        public bool IsFirstPage { get; set; }
        public bool IsLastPage { get; set; }
        public int FirstItemOnPage { get; set; }
        public int LastItemOnPage { get; set; }

        public IPagedList GetMetaData()
        {
            return this;
        }

    }

    public interface IPagedList<out T> : IPagedList, IEnumerable<T>
    {
        IPagedList GetMetaData();
        T this[int index] { get; }
        int Count { get; }
    }

    public interface IPagedList
    {
        int PageCount { get; }
        int TotalItemCount { get; }
        int PageNumber { get; }
        int PageSize { get; }
        bool HasPreviousPage { get; }
        bool HasNextPage { get; }
        bool IsFirstPage { get; }
        bool IsLastPage { get; }
        int FirstItemOnPage { get; }
        int LastItemOnPage { get; }
    }
```
Create Console Application, install AutoMapper 5.0.2, paste code and run it
You will get as a result empty list, but properties will be mapped.
It's not clear from documentation, how to map list items and properties together

**Comment by jbogard** [comment]
For future reference, you don't need to open issues for documentation. The docs are on a wiki, editable by anyone.

If there's a regression bug, open an issue for that please!

**Comment by sky-code** [comment]
Update title and message.
I Don't know this is a bug or expected behavior, because lack of documentation.

**Comment by lbargaoanu** [comment]
@jbogard This is a corner case that appeared a few times before. I think the issue should remain open for further investigation :)

**Comment by jbogard** [comment]
Agreed! @sky-code the tests are the best docs for specific behaviors, I don't have every case documented (nor will I ever, does anyone?)

**Comment by lbargaoanu** [comment]
This was difficult but [possible](https://gist.github.com/lbargaoanu/dd0d23ffe2268639d65e). Now that gist no longer works. @jbogard Do we want that gist to work again?

**Comment by sky-code** [comment]
I use custom ObjectMapper for solving this problem:
```csharp
public class PagedListObjectMapper : IObjectMapper
    {
        public object Map(ResolutionContext context)
        {
            var genericSourceType = context.SourceType.GenericTypeArguments.FirstOrDefault();
            var genericDestinationType = context.DestinationType.GenericTypeArguments.FirstOrDefault();

            var genericDestinationListType = typeof(List<>).MakeGenericType(genericDestinationType);
            var genericSourceListType = typeof(List<>).MakeGenericType(genericSourceType);
            
            var typeMap = context.ConfigurationProvider.FindTypeMapFor(genericSourceListType, genericDestinationListType);
            var destinationList = context.Engine.Map(new ResolutionContext(typeMap,
                context.SourceValue,
                context.DestinationValue,
                genericSourceListType, genericDestinationListType, new MappingOperationOptions(), Mapper.Engine));

            var newPagedListMethod = GetType()
                .GetMethod("NewPagedList", BindingFlags.NonPublic | BindingFlags.Instance)
                .MakeGenericMethod(genericSourceType, genericDestinationType);
            var result = newPagedListMethod.Invoke(this, new[] {context.SourceValue, destinationList});
            return result;
        }

        //[UsedImplicitly]
        // ReSharper disable once UnusedMember.Local
        private PagedList<TDestination> NewPagedList<TSource, TDestination>(PagedList<TSource> sourcePagedList, List<TDestination> mappedList)
        {
            return new PagedList<TDestination>(new PagingModel
            {
                Page = sourcePagedList.PageNumber,
                PageSize = sourcePagedList.PageSize
            }, sourcePagedList.TotalItemCount, mappedList);

        }

        public bool IsMatch(TypePair context)
        {
            if (!context.SourceType.IsConstructedGenericType || !context.DestinationType.IsConstructedGenericType)
            {
                return false;
            }

            var sourceGenericTypeDefinition = context.SourceType.GetGenericTypeDefinition();
            var destinationGenericTypeDefinition = context.DestinationType.GetGenericTypeDefinition();
            return sourceGenericTypeDefinition.IsAssignableFrom(typeof (PagedList<>)) &&
                   sourceGenericTypeDefinition == destinationGenericTypeDefinition;
        }
    }
```
but this code work only in AutoMapper 4.x.
I tried to update this code, but I failed.
Maybe will be better to extend AutoMapper API a little bit, add something like `ForEnumerableMember` for configuring IEnumerable source mapping

**Comment by TylerCarlson1** [comment]
@sky-code Yea 5.0 is all expression based, so it's about a hundred times more difficult to make an IObjectMapper now.

**Comment by jbogard** [comment]
@TylerCarlson1 so that thing I tried a while ago - the generic-based object mappers, that might be appropriate as a plugin mechanism.

@sky-code This code should be able to be translated to a type converter, which now support open generics. You can do `CreateMap(typeof(PagedList<>), typeof(PagedList<>)).ConvertUsing(typeof(PagedListConverter<,>))` then your converter is:

```
public class PagedListConverter<TSource, TDestination> : ITypeConverter<PagedList<TSource>, PagedList<TDestination>> {
   /// this should be way easier now
}
```

**Comment by sky-code** [comment]
Thanks for advice, I tried use ITypeConverter in AutoMapper 5.0.0, get an exception and I have forgotten about it.
this code work fine in 5.0.2 and it solve my problem:
```csharp
    public class PagedListConverter<TSource, TDestination> : ITypeConverter<PagedList<TSource>, PagedList<TDestination>>
    {
        public PagedList<TDestination> Convert(PagedList<TSource> source, PagedList<TDestination> destination,
            ResolutionContext context)
        {
            if (destination == null)
            {
                destination = new PagedList<TDestination>();
            }
            foreach (var item in source)
            {
                var dest = context.Mapper.Map<TSource, TDestination>(item);
                destination.Add(dest);
            }
            destination.PageSize = source.PageSize;
            destination.TotalItemCount = source.TotalItemCount;

            return destination;
        }
    }
```

**Comment by mikocot** [comment]
in the solution as above, we have trouble that the destination comes always as null. Even though it worked perfectly in 4.2.0 (then it was ResolutionContext.Destination). Any idea why it could happen?

**Comment by lbargaoanu** [comment]
If you don't create that list yourself and pass it in the Map call, the destination will be null. We don't pass the list created by AutoMapper.

**Comment by mikocot** [comment]
but both the source and destination list already exist, as property of the source and destination objects. What the converted is supposed to do is to merge changes between them and this is exactly how it worked in  4.2.0. The context would bring both of the list, exactly as they came in the mapped objects. Now the source is the same, but destination is null. It's same for the type converter or if you pass a single method.

**Comment by lbargaoanu** [comment]
That should work with the MyGet build.

**Comment by mikocot** [comment]
I have 5.0.2 from MyGet (installed through nuget) and it looks like even the getter of the collection is not called,

**Comment by lbargaoanu** [comment]
v5.1.0-alpha-01186 is the latest.

**Comment by mikocot** [comment]
I've seen that, but it's an alpha :) Should I understand that it's a known bug for 5.0.2 and fixed on 5.1?

**Comment by mikocot** [comment]
moving to alpha end up with stack overflow already at the configuration level :(

**Comment by lbargaoanu** [comment]
Repro?

**Comment by jbogard** [comment]
alpha because I can't pick "nightly" or "latest" it's the same as 5.0.2 but
with bugs fixed.

**Comment by mikocot** [comment]
Ok, I managed to limit the mappings and run it with the newest alpha. The bug seems to be fixed indeed, but to get there I had to skip my conditional mappings: due to StackOverflow

> config.AddConditionalObjectMapper().Where((s, d) => s.Name == d.Name + "DTO");
>  config.AddConditionalObjectMapper().Where((s, d) => s.Name + "DTO" == d.Name);

which I replaced with more generic:

> onfig.CreateMissingTypeMaps = true

but this somehow caused some other mappings that worked already with 5.0.2 and previous setup to fail with 'Cannot map XXX to Object'. I guess we'll have to move back to 4.2.1 as there seems to be too much side effects with each release and mapping rewriting.
